master/buildbot/process/builder.py

```diff
@@ -285,20 +286,40 @@ def canStartBuild(self, workerforbuilder, buildrequest):
         # check whether the locks that the build will acquire can actually be
         # acquired
         locks = self.config.locks
-        if IRenderable.providedBy(locks):
-            # collect properties that would be set for a build if we
-            # started it now and render locks using it
+        worker = workerforbuilder.worker
+        props = None
+
+        # don't unnecessarily setup properties for build
+        def setupPropsIfNeeded(props):
+            if props is not None:
+                return
             props = Properties()
             Build.setupPropertiesKnownBeforeBuildStarts(props, [buildrequest],
                                                         self, workerforbuilder)
+            return props
+
+        if isinstance(worker, AbstractLatentWorker) and \
```

Comment:
This condition looks like a code smell to me.

I would rather put this in the AbstractWorker API.

```
worker.isCompatibleWithBuild(props)
```
would always return True in the default implementation, and then subclass implemention can check for compatibility by themselves.